module/epochs/epoch_lookup.go

```diff
@@ -264,22 +263,61 @@ func (lookup *EpochLookup) handleProtocolEvents(ctx irrecoverable.SignalerContex
 		select {
 		case <-ctx.Done():
 			return
-		case block := <-lookup.committedEpochsCh:
-			epoch := lookup.state.AtBlockID(block.ID()).Epochs().Next()
-			err := lookup.cacheEpoch(epoch)
+		case processEvtFn := <-lookup.epochEvents:
+			err := processEvtFn()
 			if err != nil {
 				ctx.Throw(err)
 			}
 		}
 	}
 }
 
-// EpochCommittedPhaseStarted informs the `committee.Consensus` that the block starting the Epoch Committed Phase has been finalized.
+// EpochExtended listens to `EpochExtended` protocol notifications. The notification is queued
+// for async processing by the worker. We must process _all_ `EpochExtended` notifications.
+func (lookup *EpochLookup) EpochExtended(_ uint64, first *flow.Header, _ flow.EpochExtension) {
+	lookup.epochEvents <- func() error {
+		return lookup.processEpochExtended(first)
```

Comment:
In the Block Time Controller, we needed to query the state to re-compute the target end time and duration (not easily available just from the `EpochExtension`.

In this case though, I think we can get everything we need from the `EpochExtension` and avoid querying the state. (Although it's not strictly necessary I'd suggest passing the epoch counter as well so we can sanity check against it.)

```suggestion
func (lookup *EpochLookup) EpochExtended(epochCounter uint64, _ *flow.Header, extension flow.EpochExtension) {
	lookup.epochEvents <- func() error {
		return lookup.processEpochExtended(epochCounter, extension)
```